File: backend/internal/model/bot.go
package model

import (
	"encoding/json"
	"time"
)

// Bot status constants
const (
	BotStatusStopped  = "stopped"
	BotStatusStarting = "starting"
	BotStatusRunning  = "running"
	BotStatusError    = "error"
)

// Bot type constants
const (
	BotTypeMarketMaker = "market_maker"
	BotTypePumpHunter  = "pump_hunter"
)

// BotConfig represents a trading bot configuration
type BotConfig struct {
	ID     int64  `json:"id"`
	UserID string `json:"user_id"`
	Name   string `json:"name"`
	Type   string `json:"type"` // market_maker, pump_hunter
	Pair   string `json:"pair"`

	// Trading mode
	IsPaperTrading bool   `json:"is_paper_trading"`
	APIKeyID       *int64 `json:"api_key_id,omitempty"` // null for paper trading

	// Market Maker parameters
	InitialBalanceIDR          float64 `json:"initial_balance_idr"`
	OrderSizeIDR               float64 `json:"order_size_idr"`
	MinGapPercent              float64 `json:"min_gap_percent"`
	RepositionThresholdPercent float64 `json:"reposition_threshold_percent"`
	MaxLossIDR                 float64 `json:"max_loss_idr"`

	// Virtual balance (JSONB in DB)
	Balances map[string]float64 `json:"balances"`

	// Pump Hunter parameters
	EntryRules     *PumpHunterEntryRules     `json:"entry_rules,omitempty"`
	ExitRules      *PumpHunterExitRules      `json:"exit_rules,omitempty"`
	RiskManagement *PumpHunterRiskManagement `json:"risk_management,omitempty"`

	// Statistics
	TotalTrades    int     `json:"total_trades"`
	WinningTrades  int     `json:"winning_trades"`
	TotalProfitIDR float64 `json:"total_profit_idr"`

	// Status
	Status       string  `json:"status"` // stopped, starting, running, error
	ErrorMessage *string `json:"error_message,omitempty"`

	// Timestamps
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// BotConfigRequest represents the request to create/update a bot
type BotConfigRequest struct {
	Name           string `json:"name" binding:"required"`
	Type           string `json:"type" binding:"required,oneof=market_maker pump_hunter"`
	Pair           string `json:"pair" binding:"required_if=Type market_maker"` // Only required for market_maker
	IsPaperTrading bool   `json:"is_paper_trading"`
	APIKeyID       *int64 `json:"api_key_id"`

	// Market Maker parameters
	InitialBalanceIDR          float64 `json:"initial_balance_idr" binding:"required,gt=0"`
	OrderSizeIDR               float64 `json:"order_size_idr" binding:"required,gt=0"`
	MinGapPercent              float64 `json:"min_gap_percent" binding:"required,gte=0"`
	RepositionThresholdPercent float64 `json:"reposition_threshold_percent" binding:"required,gte=0"`
	MaxLossIDR                 float64 `json:"max_loss_idr" binding:"required_if=Type market_maker"`

	// Pump Hunter parameters
	EntryRules     *PumpHunterEntryRules     `json:"entry_rules"`
	ExitRules      *PumpHunterExitRules      `json:"exit_rules"`
	RiskManagement *PumpHunterRiskManagement `json:"risk_management"`
}

type PumpHunterEntryRules struct {
	MinPumpScore          float64  `json:"min_pump_score"`
	MinTimeframesPositive int      `json:"min_timeframes_positive"`
	Min24hVolumeIDR       float64  `json:"min_24h_volume_idr"`
	MinPriceIDR           float64  `json:"min_price_idr"`
	ExcludedPairs         []string `json:"excluded_pairs"`
	AllowedPairs          []string `json:"allowed_pairs"`
}

type PumpHunterExitRules struct {
	TargetProfitPercent    float64 `json:"target_profit_percent"`
	StopLossPercent        float64 `json:"stop_loss_percent"`
	TrailingStopEnabled    bool    `json:"trailing_stop_enabled"`
	TrailingStopPercent    float64 `json:"trailing_stop_percent"`
	MaxHoldMinutes         int     `json:"max_hold_minutes"`
	ExitOnPumpScoreDrop    bool    `json:"exit_on_pump_score_drop"`
	PumpScoreDropThreshold float64 `json:"pump_score_drop_threshold"`
}

type PumpHunterRiskManagement struct {
	MaxPositionIDR           float64 `json:"max_position_idr"`
	MaxConcurrentPositions   int     `json:"max_concurrent_positions"`
	DailyLossLimitIDR        float64 `json:"daily_loss_limit_idr"`
	CooldownAfterLossMinutes int     `json:"cooldown_after_loss_minutes"`
	MinBalanceIDR            float64 `json:"min_balance_idr"`
}

// Order represents a trading order
type Order struct {
	ID           int64   `json:"id"`
	UserID       string  `json:"user_id"`
	ParentID     int64   `json:"parent_id"`   // ID of Trade, Position, or BotConfig
	ParentType   string  `json:"parent_type"` // "trade", "position", "bot"
	OrderID      string  `json:"order_id"`    // Indodax order ID or paper order ID
	Pair         string  `json:"pair"`
	Side         string  `json:"side"`   // buy, sell
	Status       string  `json:"status"` // open, filled, cancelled
	Price        float64 `json:"price"`
	Amount       float64 `json:"amount"`
	FilledAmount float64 `json:"filled_amount"`
	IsPaperTrade bool    `json:"is_paper_trade"`

	// Timestamps
	CreatedAt time.Time  `json:"created_at"`
	UpdatedAt time.Time  `json:"updated_at"`
	FilledAt  *time.Time `json:"filled_at,omitempty"`
}

// ToJSON converts balances map to JSON string
func (b *BotConfig) ToJSON() (string, error) {
	data, err := json.Marshal(b.Balances)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

// FromJSON parses JSON string to balances map
func (b *BotConfig) FromJSON(jsonStr string) error {
	return json.Unmarshal([]byte(jsonStr), &b.Balances)
}

// GetBalance returns the balance for a specific currency
func (b *BotConfig) GetBalance(currency string) float64 {
	if b.Balances == nil {
		return 0
	}
	return b.Balances[currency]
}

// SetBalance sets the balance for a specific currency
func (b *BotConfig) SetBalance(currency string, amount float64) {
	if b.Balances == nil {
		b.Balances = make(map[string]float64)
	}
	b.Balances[currency] = amount
}

// WinRate calculates the win rate percentage
func (b *BotConfig) WinRate() float64 {
	if b.TotalTrades == 0 {
		return 0
	}
	return float64(b.WinningTrades) / float64(b.TotalTrades) * 100
}

File: frontend/src/types/bot.ts
export type BotType = 'market_maker' | 'pump_hunter';
export type BotStatus = 'stopped' | 'starting' | 'running' | 'error';

export interface PumpHunterEntryRules {
  min_pump_score: number;
  min_timeframes_positive: number;
  min_24h_volume_idr: number;
  min_price_idr: number;
  excluded_pairs: string[];
  allowed_pairs: string[];
}

export interface PumpHunterExitRules {
  target_profit_percent: number;
  stop_loss_percent: number;
  trailing_stop_enabled: boolean;
  trailing_stop_percent: number;
  max_hold_minutes: number;
  exit_on_pump_score_drop: boolean;
  pump_score_drop_threshold: number;
}

export interface PumpHunterRiskManagement {
  max_position_idr: number;
  max_concurrent_positions: number;
  daily_loss_limit_idr: number;
  cooldown_after_loss_minutes: number;
  min_balance_idr: number;
}

export interface BotConfig {
  id: number;
  user_id: string;
  name: string;
  type: BotType;
  pair: string;
  is_paper_trading: boolean;
  api_key_id?: number | null;

  // Market Maker parameters
  initial_balance_idr: number;
  order_size_idr: number;
  min_gap_percent: number;
  reposition_threshold_percent: number;
  max_loss_idr: number;

  // Virtual balances
  balances: Record<string, number>;

  // Pump Hunter parameters
  entry_rules?: PumpHunterEntryRules;
  exit_rules?: PumpHunterExitRules;
  risk_management?: PumpHunterRiskManagement;

  // Statistics
  total_trades: number;
  winning_trades: number;
  total_profit_idr: number;
  win_rate?: number; // Calculated field for UI

  // Status
  status: BotStatus;
  error_message?: string;

  created_at: string;
  updated_at: string;
}

export interface BotConfigRequest {
  name: string;
  type: BotType;
  pair: string;
  is_paper_trading: boolean;
  api_key_id?: number | null;
  initial_balance_idr: number;
  order_size_idr: number;
  min_gap_percent: number;
  reposition_threshold_percent: number;
  max_loss_idr: number;
  entry_rules?: PumpHunterEntryRules;
  exit_rules?: PumpHunterExitRules;
  risk_management?: PumpHunterRiskManagement;
}

export interface UpdateBotRequest extends Partial<BotConfigRequest> {}

export interface Position {
  id: number;
  bot_config_id: number;
  user_id: string;
  pair: string;
  status: 'open' | 'closed';
  entry_price: number;
  entry_quantity: number;
  entry_amount_idr: number;
  entry_order_id?: string;
  entry_pump_score?: number;
  entry_at: string;
  exit_price?: number;
  exit_quantity?: number;
  exit_amount_idr?: number;
  exit_order_id?: string;
  exit_at?: string;
  highest_price?: number;
  lowest_price?: number;
  profit_idr: number;
  profit_percent: number;
  close_reason?: string;
  is_paper_trade: boolean;
  created_at: string;
  updated_at: string;
  // Legacy fields for backward compatibility
  bot_id?: number; // Maps to bot_config_id
  entry_amount?: number; // Maps to entry_amount_idr
  exit_amount?: number; // Maps to exit_amount_idr
  closed_at?: string; // Maps to exit_at
}

export interface Order {
  id: number;
  user_id: string;
  parent_id: number;
  parent_type: 'bot' | 'trade' | 'position';
  order_id: string; // Indodax order ID
  pair: string;
  side: 'buy' | 'sell';
  status: 'pending' | 'open' | 'filled' | 'cancelled' | 'failed' | 'stopped';
  price: number;
  amount: number;
  filled_amount: number;
  is_paper_trade: boolean;
  created_at: string;
  updated_at: string;
  filled_at?: string;
}

export interface BotSummary {
  bot_id: number;
  type: string;
  status: string;
  pair?: string;
  total_trades: number;
  winning_trades: number;
  losing_trades: number;
  win_rate: number;
  total_profit_idr: number;
  average_profit: number;
  uptime: string;
  last_trade_at?: string;
  balance_idr?: number;
  balance_coin?: number;
  balance_coin_name?: string;
  open_positions?: number; // Added for convenience info
  profit_factor?: number; // Added for convenience info
}
